docs/configuration.md
# Configuration

There are three different things you can configure within Mquery: core, plugins
and ursadb. Unfortunately, all are configured differently.

## Mquery core configuration

Mquery is configured with [typed-config](https://github.com/bwindsor/typed-config). There are two ways to pass every configuration field - with
a config file, or a environment variable. For example:

```ini
[redis]
host=redis-server.example.com

[mquery]
backend=tcp://ursadb-server.example.com:9281
plugins=plugins.archive:GzipPlugin
```

This is a simple INI configuration file that mquery understands. It should
be saved in a file called `mquery.ini`. The file should be in one of the
following locations (checked in that order):

* Mquery's working directory (usually `src` folder in the cloned repository)
* In the current user's xdg config directory: `~/.config/mquery/mquery.ini`
* In the system config directory: `/etc/mquery/mquery.ini`

Alternatively, you can use environment variables to configure mquery. All
field names are mapped intuitively to environment variables by joining
the ini section name with a key name - for example, to change redis host
value use `REDIS_HOST`. Environment variables take precedence over values
from the config file!

Currently, supported configuration keys are:

- `redis.host`: Hostname of a main redis server.
- `redis.port`: Port of a main redis server.
- `mquery.backend`: URL to a ursadb instance (for example,
  `tcp://ursadb-server:9281`)
- `mquery.plugins`: List of supported plugins, separated by commas (for
  example `plugins.archive:GzipPlugin, plugins.custom:CustomPlugin`)
- `mquery.yara_limit`: Maximum number of yara-scanned files per query (0 means no limit).
- `mquery.about`: HTML code to be displayed on the about page.
- `rq.job_timeout`: Timeout value for rq jobs. The default is OK for most users.

## Mquery plugin configuration

In contrast to the core configuration, plugins can be configured dynamically.
Every worker registers its list of active plugins in the database, and it's
possible to configure them using the web UI:

![](./plugin-config.png)

This configuration mechanism is used by the plugins shipped with Mquery.
Despite this, it's optional, and plugin authors don't have to use it.
Since plugins are arbitrary code, plugins can read their configuration from
anywhere they want, including the environment, other config files, etc.

It's also easy to use the same config file for Mquery and plugins - see
[example_typed_config_plugin.py](../src/plugins/example_typed_config_plugin.py)
file for an example.

## UrsaDB configuration

UrsaDB is not technically part of Mquery, but both systems work closely
together and depend on each other for optimal performance.

Mquery currently does not allow you to configure UrsaDB nicely.
You have to do it "manually", by connecting with `ursacli` program to the
TCP port exposed by UrsaDB. This program is built together with UrsaDB, and
available in all official docker images. You can execute it in docker compose
like this:

```
docker compose exec ursadb ursacli
```

Or you can download the latest ursadb release and run a client from there.

To set a configuration field, issue a command like this:

```
$ ursacli
ursadb> config set "database_workers" 10;
```

The configuration keys are already documented in the UrsaDB's docs here:
https://cert-polska.github.io/ursadb/configuration.html. We won't copy
all relevant information here, but the most important config keys are:

* `database_workers` - How many tasks can be processed at once (you can
  increase this for strong servers, restart the database to apply).
* `merge_max_files` - Biggest supported dataset size. UrsaDB keeps indexed
  files in so-called "datasets". The fewer datasets the faster the database is,
  but you might not want overly huge datasets (at some point merging datasets
  becomes very slow, and you get diminishing returns for merging them). Decide
  on the value before indexing your files. Good values include the default
  (infinite), 10 million, and 1 million.
* `merge_max_datasets` - a very memory and CPU-intensive operation.
  If your database OOMs during indexing, consider lowering this number and
  the number of `database_workers` that are merging in parallel.

## .env file

Finally, in the main directory of the repository there is a file named `.env`.
Mquery does not use it in any way, but it's read by Docker.

```bash
$ cat .env
# This file is only relevant for docker compose deployments.

# Directory where your samples are stored. By default you have to copy them
# to ./samples subdirectory in this repository.
SAMPLES_DIR=./samples
# Directory where the index files should be saved. By default ./index
# subdirectory in this repository.
INDEX_DIR=./index
```

If you use docker compose to start mquery, you can use this file to specify
a location on the host for your samples_dir and index_dir. These variables are
then used when creating containers. See for example ursadb container spec:

```yaml
  ursadb:
    restart: always
    image: mqueryci/ursadb:v1.5.0
    ports:
    - "127.0.0.1:9281:9281"
    volumes:
    - "${SAMPLES_DIR}:/mnt/samples"
    - "${INDEX_DIR}:/var/lib/ursadb"
    user: "0:0"
```

As you can see, variables from `.env` are used to specify mount point for
the data volumes. You can also ignore this file, and edit docker compose
directly to your liking.


RELEASES.md
# Version 1.4

### Breaking changes

**[Breaking change]** mquery now users typedconfig library instead of the previous config.py file.

 - if you deployed mquery using docker (configurable by environment variables)
   then no action is required and this is backwards-compatible for you
 - if you deployed mquery natively using the default configuration, no action is required
 - finally, if you deploy mquery natively and changed the default config.py, you will have to create a mquery.ini
   file with your config. The format is very simple. Example of a complete config file (there are only 4 possible
   configuration keys supported currently. All are optional):

```
[redis]
host=localhost
port=6379

[mquery]
backend=tcp://localhost:9281
plugins=
```

### New features

- It's now possible to limit the number of YARA-scanned files (#339)
- It's now possible to disallow running slow queries (#315, #312)
- Added a configurable /about page, to describe your instance (#341)
- Daemon now has a --scale flag, to automatically fork into mutliple processes (#298)
- More flexible user roles (#350, #314)

### Documentation

- Mquery component documentation (#334)
- YARA support documentation (#333)
- S3 support documentation (#327)

### UI Improvements

- Progress bar now shows more information (#345)
- Counter race condition fixed (#348)
- Bootstrap update and following fixes (#346, 

### Improvements

- A big backend improvement - jobs are now scheduled using the RQ framework (#317)
- Exceptions thrown during filtering with plugins are now handled correctly (#317)
- Login is now faster - there are no unnecessary redirects (#322)

### Bugfixes

- /about route fixed (#343)
- Indexing script won't skip the last few files anymore (#328)
- Actually raise errors from the API (#311)
- Fix multi-agent job completion (#282)

### Others

- Dockerignore and Gitignore updated (#344)
- Some obsolete features removed from the codebase (#330, #313, #306)


# Version 1.3

### New features

- User accounts with OIDC (#250, #251, #252, #253, #255, #258, #266, #265, #274, #276, #278, #280)

### UI Improvements

- Multiselect for sample tags [(#164)](https://github.com/CERT-Polska/mquery/pull/164)
- Ctrl+enter now submits a job in the query window [(#217)](https://github.com/CERT-Polska/mquery/pull/217)
- Added a button to copy all matched hashes [(#239)](https://github.com/CERT-Polska/mquery/pull/239)

### Improvements

- A bit better support for Yara rules:
    - Improve parsing of string count expressions [(#269)](https://github.com/CERT-Polska/mquery/pull/269)
    - Discard partial "or" expressions [(#190)](https://github.com/CERT-Polska/mquery/pull/190)
    - Fixed the regex parsing [(#229)](https://github.com/CERT-Polska/mquery/pull/229)

### Bugfixes

- NPM made a breaking change that broke our builds - fixed with [#272](https://github.com/CERT-Polska/mquery/pull/66)
- Add a missing /config route [#209](https://github.com/CERT-Polska/mquery/pull/209)

### Others

- Remove ursadb repository as a submodule [(#277)](https://github.com/CERT-Polska/mquery/pull/277)
- Automatically build and push docker images on merge [(#262)](https://github.com/CERT-Polska/mquery/pull/262)
- Various refactoring changes, like [(#199)](https://github.com/CERT-Polska/mquery/pull/199) or #245

# Version 1.2

![](./docs/interface-v1.2.gif)

### New features

- Results streaming (first results appear faster) [(#59)](https://github.com/CERT-Polska/mquery/pull/59)
- Support for distributed Ursadb in the backend [(#119)](https://github.com/CERT-Polska/mquery/pull/119)
- Powerful plugin support (See the [documentation](./docs/plugins.md))
    - Configurable from the web UI [(#132)](https://github.com/CERT-Polska/mquery/pull/132)
- Support for Ursadb tags (for example, to tag collections as benign or malicious) [(#44)](https://github.com/CERT-Polska/mquery/pull/44)

### UI Improvements

- Use Monaco IDE as yara editor [(#109)](https://github.com/CERT-Polska/mquery/pull/#109)
- Results view improvements [issue #82](https://github.com/CERT-Polska/mquery/issues/82)
    - Remove old jobs [(#180)](https://github.com/CERT-Polska/mquery/pull/#180)
    - Show query ETA and duration [(#175)](https://github.com/CERT-Polska/mquery/pull/175)
    - Show sample sha256 [(#156)](https://github.com/CERT-Polska/mquery/pull/156) [(#167)](https://github.com/CERT-Polska/mquery/pull/167)
    - Download matched files or sha256 hashes [(#176)](https://github.com/CERT-Polska/mquery/pull/#176) [(#163)](https://github.com/CERT-Polska/mquery/pull/#163)
    - Add pagination to the results table [(#96)](https://github.com/CERT-Polska/mquery/pull/#96)
- Status page improvements:
    - Display file count along with dataset size [(#91)](https://github.com/CERT-Polska/mquery/pull/#91)
- Filter jobs by author, status and others [(#152)](https://github.com/CERT-Polska/mquery/pull/#152)
- Show number of errors (for example, missing files) per job [(#148)](https://github.com/CERT-Polska/mquery/pull/#148)

### Improvements

- Much better Yara support [(issue #41)](https://github.com/CERT-Polska/mquery/issues/41):
    - Multiple rules in a query [(#55)](https://github.com/CERT-Polska/mquery/pull/55)
    - Private and global rules [(#55)](https://github.com/CERT-Polska/mquery/pull/55)
    - Case insensitive strings (`nocase` modifier) [(#136)](https://github.com/CERT-Polska/mquery/pull/136)
    - Regexes [(#169)](https://github.com/CERT-Polska/mquery/pull/169)
    - Strings that are both `ascii` and `wide` [(#65)](https://github.com/CERT-Polska/mquery/pull/65)
    - `xor` modifier (without ranges) [(#98)](https://github.com/CERT-Polska/mquery/pull/98)
    - Anonymous variables [(#66)](https://github.com/CERT-Polska/mquery/pull/66)
- Use FastAPI framework, instead of Flask to improve performance and get API documentation for free
- Document the API and add swagger UI to `/docs` endpoint
- Cache parsed Yara rules
- Batch files when matching yara rules, to improve performance

### Others

src/config.py
from typedconfig import Config, key, section, group_key  # type: ignore
from typedconfig.source import EnvironmentConfigSource, IniFileConfigSource  # type: ignore
import os


@section("redis")
class RedisConfig(Config):
    # Hostname of a configured redis instance.
    host = key(cast=str, required=False, default="localhost")
    # Port of a configured redis instance.
    port = key(cast=int, required=False, default=6379)


@section("database")
class DatabaseConfig(Config):
    # URL of a configured sql database.
    url = key(
        cast=str, required=False, default="postgresql://localhost:5432/mquery"
    )


@section("rq")
class RqConfig(Config):
    # Timeout value for rq jobs.
    job_timeout = key(cast=int, required=False, default=300)


@section("mquery")
class MqueryConfig(Config):
    # URL to a UrsaDB instance.
    backend = key(cast=str, required=False, default="tcp://127.0.0.1:9281")
    # List of plugin specifications separated by comma, for example
    # "plugins.archive:GzipPlugin, plugins.custom:CustomPlugin"
    plugins = key(cast=str, required=False, default="")
    # Maximum number of yara-scanned files per query (0 means no limit).
    yara_limit = key(cast=int, required=False, default=0)
    # Html code to be displayed on the about page.
    about = key(cast=str, required=False, default="")


class AppConfig(Config):
    redis = group_key(RedisConfig)
    database = group_key(DatabaseConfig)
    rq = group_key(RqConfig)
    mquery = group_key(MqueryConfig)


def _config_sources():
    return [
        EnvironmentConfigSource(),
        IniFileConfigSource("mquery.ini", must_exist=False),
        IniFileConfigSource(
            os.path.expanduser("~/.config/mquery/mquery.ini"), must_exist=False
        ),
        IniFileConfigSource("/etc/mquery/mquery.ini", must_exist=False),
    ]


app_config = AppConfig(sources=_config_sources())
